dht_bootstrap.go

```diff
@@ -41,52 +43,71 @@ func init() {
 	}
 }
 
-// BootstrapConfig runs cfg.Queries bootstrap queries every cfg.BucketPeriod.
-func (dht *IpfsDHT) Bootstrap(ctx context.Context) error {
-	triggerBootstrapFnc := func() {
-		logger.Infof("triggerBootstrapFnc: RT only has %d peers which is less than the min threshold of %d, triggering self & bucket bootstrap",
-			dht.routingTable.Size(), minRTBootstrapThreshold)
+type bootstrapReq struct {
+	errChan chan error
+}
 
-		if err := dht.selfWalk(ctx); err != nil {
-			logger.Warningf("triggerBootstrapFnc: self walk: error: %s", err)
-		}
+func makeBootstrapReq() *bootstrapReq {
+	errChan := make(chan error, 1)
+	return &bootstrapReq{errChan}
+}
 
-		if err := dht.bootstrapBuckets(ctx); err != nil {
-			logger.Warningf("triggerBootstrapFnc: bootstrap buckets: error bootstrapping: %s", err)
-		}
-	}
+// Bootstrap  i
+func (dht *IpfsDHT) startBootstrapping() error {
+	// scan the RT table periodically & do a random walk on k-buckets that haven't been queried since the given bucket period
+	dht.proc.Go(func(proc process.Process) {
+		ctx := processctx.OnClosingContext(proc)
 
-	// we should query for self periodically so we can discover closer peers
-	go func() {
-		for {
-			err := dht.selfWalk(ctx)
-			if err != nil {
-				logger.Warningf("self walk: error: %s", err)
-			}
-			select {
-			case <-time.After(dht.bootstrapCfg.SelfQueryInterval):
-			case <-ctx.Done():
-				return
+		scanInterval := time.NewTicker(dht.bootstrapCfg.BucketPeriod)
+		defer scanInterval.Stop()
+
+		var (
+			lastSelfWalk time.Time
+		)
+
+		// run bootstrap if option is set
+		if dht.triggerAutoBootstrap {
+			if err := dht.doBootstrap(ctx, true, &lastSelfWalk); err != nil {
+				logger.Warningf("bootstrap error: %s", err)
 			}
 		}
-	}()
 
-	// scan the RT table periodically & do a random walk on k-buckets that haven't been queried since the given bucket period
-	go func() {
 		for {
-			err := dht.bootstrapBuckets(ctx)
-			if err != nil {
-				logger.Warningf("bootstrap buckets: error bootstrapping: %s", err)
-			}
 			select {
-			case <-time.After(dht.bootstrapCfg.RoutingTableScanInterval):
-			case <-dht.triggerBootstrap:
-				triggerBootstrapFnc()
+			case now := <-scanInterval.C:
+				walkSelf := now.After(lastSelfWalk.Add(dht.bootstrapCfg.SelfQueryInterval))
+				if err := dht.doBootstrap(ctx, walkSelf, &lastSelfWalk); err != nil {
+					logger.Warning("bootstrap error: %s", err)
+				}
+			case req := <-dht.triggerBootstrap:
+				logger.Infof("triggering a bootstrap: RT has %d peers", dht.routingTable.Size())
+				err := dht.doBootstrap(ctx, true, &lastSelfWalk)
+				select {
```

Comment:
If the err channel always has 1 slot, we shouldn't need to do this. We should be able to just run:

```go
req.errChan <- err
close(req.errChan)
```

Reply:
@stebalien: Oh ofcourse, but we can't rely on clients always sending a buffered channel.

Reply:
Isn't this internal?

Reply:
@Stebalien  

It really is just a precautionary measure. If in the future, a go-routine within Dht("client") decides to a trigger a bootstrap & passes an unbuffered channel because of whatever reason(programmer oversight for example), the whole bootstrapping loop comes to a halt unless the "client"  reads from it. We don't want to rely on that.

Reply:
Meh, either way. IMO, the idiom of passing respone channels with a size of 1 is common enough that users should always check how a channel is being used before passing an unbuffered response channel.

But either way works.